scheduler/src/cook/mesos/constraints.clj

```diff
@@ -116,7 +117,37 @@
   (let [needs-gpus? (job-needs-gpus? job)]
     (->gpu-host-constraint job needs-gpus?)))
 
-(def job-constraint-constructors [build-novel-host-constraint build-gpu-host-constraint])
+(defrecord user-defined-constraint [constraints]
+  JobConstraint
+  (job-constraint-name [this] (get-class-name this))
+  (job-constraint-evaluate
+    [this _ vm-attributes]
+    (let [vm-passes-constraint?
+          (fn vm-passes-constraint? [{attribute :constraint/attribute
+                                      pattern :constraint/pattern
+                                      operator :constraint/operator}]
+            (let [vm-attribute-value (get vm-attributes attribute)]
+              (condp = operator
+                :constraint.operator/equals (= pattern vm-attribute-value)
+                :else (do
+                        (log/error (str "Unknown operator " operator
+                                        " api.clj should have prevented this from happening."))
+                        true))))
+          constraint->passes? (pc/map-from-keys vm-passes-constraint? constraints)
```

Comment:
Good point, I was going to say which constraints failed but decided against it because it causes a lot of memory allocs.